File: daemon/__tests__/migrate.test.ts
import { mkdir, readFile, readdir, stat, writeFile } from 'node:fs/promises'
import path from 'node:path'
import { afterAll, describe, expect, it } from 'vitest'
import type { BroodmotherConfig } from '@broodmother/types/config'
import { defaultConfig } from '@broodmother/config'
import { migrate } from '../src/migrate'
import { listProfiles } from '../src/profiles'
import { listRepos } from '@broodmother/repo'
import { PRIMARY } from '@broodmother/branch'
import { cleanup, git, tempDir } from '../src/test'

afterAll(cleanup)

const names = async (dir: string) =>
  (await readdir(dir, { withFileTypes: true }))
    .filter((entry) => entry.isDirectory())
    .map((entry) => entry.name)
    .sort()

const loaded = (config: BroodmotherConfig, bindings: Record<string, string> = {}) => ({
  config,
  reset: [],
  bindings,
})

/** A project the way it was before checkouts: the folder is the checkout. */
async function project(home: string, name: string) {
  const dir = path.join(home, name)
  await mkdir(path.join(dir, '.git'), { recursive: true })
  await writeFile(path.join(dir, '.git', 'HEAD'), 'ref: refs/heads/main\n')
  await writeFile(path.join(dir, 'README.md'), `# ${name}\n`)
  await mkdir(path.join(dir, 'Notes'), { recursive: true })
  return dir
}

/** A profile the way it was: a file in `profiles/`. */
async function profile(home: string, name: string, identity: object = {}) {
  await mkdir(path.join(home, 'profiles'), { recursive: true })
  const file = path.join(home, 'profiles', `${name}.json`)
  await writeFile(file, JSON.stringify(identity))
  return file
}

/** A repository outside the home, the way a repo used to be linked. */
async function legacyRepo(name: string) {
  const dir = path.join(await tempDir(), name)
  await mkdir(dir, { recursive: true })
  await git(dir, 'init', '--initial-branch=main')
  await writeFile(path.join(dir, 'main.rs'), 'fn main() {}\n')
  await git(dir, 'add', '-A')
  await git(dir, 'commit', '-m', 'init')
  return dir
}

async function register(dir: string, entries: Record<string, string>) {
  await mkdir(path.join(dir, '.projects'), { recursive: true })
  await writeFile(path.join(dir, '.projects', 'projects.json'), JSON.stringify(entries))
}

describe('migrate', () => {
  it('puts each project inside the profile it was bound to, as its local', async () => {
    const home = await tempDir()
    await profile(home, 'ada')
    const was = await project(home, 'handbook')

    const result = await migrate(home, loaded(defaultConfig(was), { [was]: 'ada' }))

    const now = path.join(home, 'ada', 'handbook')
    expect(result.moved).toEqual([now])
    expect(await names(home)).toEqual(['ada'])
    expect(await names(now)).toEqual([PRIMARY])
    expect(await readFile(path.join(now, PRIMARY, 'README.md'), 'utf8')).toBe(
      '# handbook\n',
    )
    // `.git` moved with everything else, which is what makes the folder that arrived a
    // repository rather than a copy of one.
    expect(await names(path.join(now, PRIMARY))).toEqual(['.git', 'Notes'])
    expect(result.config.projectPath).toBe(now)
    expect(result.config.profile).toBe('ada')
  })

  it('makes the profiles folder into the folders the profiles now are', async () => {
    const home = await tempDir()
    await profile(home, 'ada', { color: '#abcdef' })

    await migrate(home, loaded(defaultConfig(null)))

    const profiles = await listProfiles(home)
    expect(profiles.map((one) => one.name)).toEqual(['ada'])
    expect(profiles[0]?.color).toBe('#abcdef')
    expect(profiles[0]?.path).toBe(path.join(home, 'ada', 'profile.json'))
    await expect(stat(path.join(home, 'profiles'))).rejects.toThrow()
  })

  /* The key moved with the profile, and the profile named it by absolute path. */
  it('repoints a profile at the key that moved with it', async () => {
    const home = await tempDir()
    const was = path.join(home, 'profiles', 'ada.key')
    await profile(home, 'ada', { sshKeyPath: was })
    await writeFile(was, 'private\n')
    await writeFile(`${was}.pub`, 'ssh-ed25519 AAAA ada\n')

    await migrate(home, loaded(defaultConfig(null)))

    const key = path.join(home, 'ada', 'profile.key')
    expect((await listProfiles(home))[0]?.sshKeyPath).toBe(key)
    expect(await readFile(`${key}.pub`, 'utf8')).toBe('ssh-ed25519 AAAA ada\n')
  })

  it('gives a project nobody bound to the first profile there is', async () => {
    const home = await tempDir()
    await profile(home, 'ada')
    await project(home, 'orphan')

    await migrate(home, loaded(defaultConfig(null)))

    expect(await names(path.join(home, 'ada'))).toEqual(['orphan'])
  })

  /* A folder dropped into the home by hand is a project, and it was never bound to anyone. */
  it('makes a profile to hold the projects of a home that had none', async () => {
    const home = await tempDir()
    await project(home, 'handbook')

    const result = await migrate(home, loaded(defaultConfig(null)))

    expect(await names(home)).toEqual(['default'])
    expect((await listProfiles(home)).map((one) => one.name)).toEqual(['default'])
    expect(result.config.profile).toBe('default')
  })

  it('moves every repository the registry pointed at into its project', async () => {
    const home = await tempDir()
    await profile(home, 'ada')
    const was = await project(home, 'handbook')
    const api = await legacyRepo('api')
    await register(was, { api })

    await migrate(home, loaded(defaultConfig(was), { [was]: 'ada' }))

    const now = path.join(home, 'ada', 'handbook')
    const [repo] = await listRepos(now)
    expect(repo).toEqual({
      name: 'api',
      repo: path.join(now, '.repos', 'api', PRIMARY),
    })
    expect(await readFile(path.join(repo!.repo, 'main.rs'), 'utf8')).toBe(
      'fn main() {}\n',
    )
    await expect(stat(api)).rejects.toThrow()
    // The registry has nothing left to say: a repo is a folder now.
    await expect(stat(path.join(now, '.repos', 'projects.json'))).rejects.toThrow()
    await expect(stat(path.join(now, '.projects'))).rejects.toThrow()
  })

  /* A worktree records where its repository is, and the repository records where the
     worktree is — both in absolute paths that the move invalidates. */
  it('repairs the checkouts of a repository it moved', async () => {
    const home = await tempDir()
    await profile(home, 'ada')
    const was = await project(home, 'handbook')
    const api = await legacyRepo('api')
    await git(api, 'branch', 'fix-login')
    await mkdir(path.join(was, '.projects', 'api'), { recursive: true })
    await git(api, 'worktree', 'add', path.join(was, '.projects', 'api', 'fix-login'))
    await register(was, { api })

    await migrate(home, loaded(defaultConfig(was), { [was]: 'ada' }))

    const now = path.join(home, 'ada', 'handbook', '.repos', 'api')
    const listed = await git(path.join(now, PRIMARY), 'worktree', 'list')
    expect(listed.stdout).toContain(path.join(now, 'fix-login'))
    // Answering from inside the worktree is the other direction of the same link.
    const branch = await git(path.join(now, 'fix-login'), 'branch', '--show-current')
    expect(branch.stdout.trim()).toBe('fix-login')
  })

  /* Repos were projects, and a project already in the new layout keeps them under the old
     name of the folder — the only thing this migration touches in it. */
  it('renames .projects to .repos in a project and repairs the checkouts inside', async () => {
    const home = await tempDir()
    // A profile and a project already in the new layout: a folder, with the file that makes it one.
    await mkdir(path.join(home, 'ada'), { recursive: true })
    await writeFile(path.join(home, 'ada', 'profile.json'), '{}\n')
    const projectDir = path.join(home, 'ada', 'handbook')
    await mkdir(path.join(projectDir, PRIMARY), { recursive: true })
    const api = path.join(projectDir, '.projects', 'api', PRIMARY)
    await mkdir(api, { recursive: true })
    await git(api, 'init', '--initial-branch=main')
    await writeFile(path.join(api, 'main.rs'), 'fn main() {}\n')
    await git(api, 'add', '-A')
    await git(api, 'commit', '-m', 'init')
    await git(api, 'branch', 'fix-login')
    await git(api, 'worktree', 'add', path.join(projectDir, '.projects', 'api', 'fix-login'))

    await migrate(home, loaded(defaultConfig(projectDir)))

    const now = path.join(projectDir, '.repos', 'api')
    expect(await listRepos(projectDir)).toEqual([
      { name: 'api', repo: path.join(now, PRIMARY) },
    ])
    await expect(stat(path.join(projectDir, '.projects'))).rejects.toThrow()
    const listed = await git(path.join(now, PRIMARY), 'worktree', 'list')
    expect(listed.stdout).toContain(path.join(now, 'fix-login'))
    const back = await git(path.join(now, 'fix-login'), 'rev-parse', '--git-common-dir')
    expect(path.resolve(path.join(now, 'fix-login'), back.stdout.trim())).toBe(
      path.join(now, PRIMARY, '.git'),
    )
    expect(await names(projectDir)).toEqual(['.repos', PRIMARY])
  })

  it('moves everything this machine filed under the project onto its new path', async () => {
    const home = await tempDir()
    await profile(home, 'ada')
    const was = await project(home, 'handbook')
    const config: BroodmotherConfig = {
      ...defaultConfig(was),
      checkouts: { [was]: 'feat-sync' },
      repo: { [was]: 'api' },
      repoBranch: { [`${was}#api`]: 'fix-login' },
    }

    const result = await migrate(home, loaded(config, { [was]: 'ada' }))

    const now = path.join(home, 'ada', 'handbook')
    expect(result.config.checkouts).toEqual({ [now]: 'feat-sync' })
    expect(result.config.repo).toEqual({ [now]: 'api' })
    expect(result.config.repoBranch).toEqual({ [`${now}#api`]: 'fix-login' })
  })

  it('is a no-op the second time', async () => {
    const home = await tempDir()
    await profile(home, 'ada')
    const was = await project(home, 'handbook')

    const once = await migrate(home, loaded(defaultConfig(was), { [was]: 'ada' }))
    const twice = await migrate(home, loaded(once.config))

    expect(twice.moved).toEqual([])
    expect(twice.config).toEqual(once.config)
    expect(await names(path.join(home, 'ada', 'handbook'))).toEqual([PRIMARY])
  })

  it('leaves a home that has nothing in it alone', async () => {
    const home = await tempDir()

    const result = await migrate(home, loaded(defaultConfig(null)))

    expect(result.moved).toEqual([])
    expect(await names(home)).toEqual([])
  })
})

File: daemon-go/tests/migrate/migrate_test.go
package migrate_test

import (
	. "github.com/broodmotherai/broodmother/daemon-go/internal/migrate"

	"encoding/json"
	"io/fs"
	"os"
	"path/filepath"
	"slices"
	"strings"
	"testing"

	"github.com/broodmotherai/broodmother/daemon-go/internal/config"
)

// build lays out a home. A name ending in `/` is an empty folder; everything else is a file with
// the bytes given, and the folders above it are made.
func build(t *testing.T, files map[string]string) string {
	t.Helper()
	home := t.TempDir()
	for name, body := range files {
		full := filepath.Join(home, filepath.FromSlash(name))
		if strings.HasSuffix(name, "/") {
			if err := os.MkdirAll(full, 0o755); err != nil {
				t.Fatal(err)
			}
			continue
		}
		if err := os.MkdirAll(filepath.Dir(full), 0o755); err != nil {
			t.Fatal(err)
		}
		if err := os.WriteFile(full, []byte(body), 0o644); err != nil {
			t.Fatal(err)
		}
	}
	return home
}

// tree is every path under the home, relative and slash-separated, so a whole layout is one
// comparison rather than a dozen stats.
func tree(t *testing.T, home string) []string {
	t.Helper()
	var found []string
	err := filepath.WalkDir(home, func(path string, entry fs.DirEntry, err error) error {
		if err != nil || path == home {
			return err
		}
		rest, err := filepath.Rel(home, path)
		if err != nil {
			return err
		}
		name := filepath.ToSlash(rest)
		if entry.IsDir() {
			name += "/"
		}
		found = append(found, name)
		return nil
	})
	if err != nil {
		t.Fatal(err)
	}
	slices.Sort(found)
	return found
}

func run(t *testing.T, home string, loaded config.Loaded) config.Config {
	t.Helper()
	held, err := Run(home, loaded)
	if err != nil {
		t.Fatalf("migration: %v", err)
	}
	return held
}

func plain() config.Loaded {
	return config.Loaded{Config: config.Default(nil), Bindings: map[string]string{}}
}

func has(t *testing.T, home string, paths ...string) {
	t.Helper()
	found := tree(t, home)
	for _, path := range paths {
		if !slices.Contains(found, path) {
			t.Errorf("no %s — the home is %v", path, found)
		}
	}
}

func hasNo(t *testing.T, home string, paths ...string) {
	t.Helper()
	found := tree(t, home)
	for _, path := range paths {
		if slices.Contains(found, path) {
			t.Errorf("%s is still there — the home is %v", path, found)
		}
	}
}

// The whole cost of running on a home that needs nothing: a handful of directory reads and not
// one write. A migration that moved something here would move it on every start.
func TestLeavesAHomeAlreadyInShapeAlone(t *testing.T) {
	home := build(t, map[string]string{
		"ada/profile.json":                   "{}\n",
		"ada/notes/local/one.md":             "hello\n",
		"ada/notes/.repos/x/local/README.md": "x\n",
	})
	before := tree(t, home)
	held := run(t, home, plain())
	if got := tree(t, home); !slices.Equal(got, before) {
		t.Errorf("moved something:\n got %v\nwant %v", got, before)
	}
	if held.Profile == nil || *held.Profile != "ada" {
		t.Errorf("profile is %v", held.Profile)
	}
}

// A home from before profiles existed: every folder in it is a project, and there is no profile
// to hand them to.
func TestMakesAProfileForAHomeThatNeverHadOne(t *testing.T) {
	home := build(t, map[string]string{"notes/one.md": "hello\n"})
	held := run(t, home, plain())

	has(t, home, "default/", "default/profile.json", "default/notes/", "default/notes/local/", "default/notes/local/one.md")
	hasNo(t, home, "notes/", ".migrating/")
	if held.Profile == nil || *held.Profile != "default" {
		t.Errorf("profile is %v", held.Profile)
	}
}

// The binding was a map in the config before a project sat inside the profile it commits as.
func TestHandsEachProjectToTheProfileItWasBoundTo(t *testing.T) {
	home := build(t, map[string]string{
		"ada/profile.json": "{}\n",
		"bo/profile.json":  "{}\n",
		"notes/one.md":     "hello\n",
		"drafts/two.md":    "hello\n",
	})
	loaded := plain()
	loaded.Bindings = map[string]string{filepath.Join(home, "notes"): "bo"}
	run(t, home, loaded)

	// Bound to bo; unbound, so the first profile there is.
	has(t, home, "bo/notes/local/one.md", "ada/drafts/local/two.md")
}

// `profiles/ada.json` and the key beside it become the folder `ada/` holds.
func TestAProfileFileBecomesAFolder(t *testing.T) {
	home := build(t, map[string]string{
		"profiles/ada.json":    `{"color":"#8fb8d8","sshKeyPath":"` + "SSHKEY" + `"}` + "\n",
		"profiles/ada.key":     "private\n",
		"profiles/ada.key.pub": "public\n",
	})
	// The path is only known once the home is, and the repoint turns on it being exact.
	file := filepath.Join(home, "profiles", "ada.json")
	body, err := os.ReadFile(file)
	if err != nil {
		t.Fatal(err)
	}
	rewritten := strings.Replace(string(body), "SSHKEY", filepath.Join(home, "profiles", "ada.key"), 1)
	if err := os.WriteFile(file, []byte(rewritten), 0o644); err != nil {
		t.Fatal(err)
	}

	run(t, home, plain())
	has(t, home, "ada/", "ada/profile.json", "ada/profile.key", "ada/profile.key.pub")
	hasNo(t, home, "profiles/")

	var held map[string]any
	saved, err := os.ReadFile(filepath.Join(home, "ada", "profile.json"))
	if err != nil {
		t.Fatal(err)
	}
	if err := json.Unmarshal(saved, &held); err != nil {
		t.Fatal(err)
	}
	if want := filepath.Join(home, "ada", "profile.key"); held["sshKeyPath"] != want {
		t.Errorf("key is %v, want %s", held["sshKeyPath"], want)
	}
	if held["color"] != "#8fb8d8" {
		t.Errorf("lost the rest of the file: %v", held)
	}
}

// A profile and a project could share a name, and the profile is the one that keeps it: the
// projects are staged out of the way before any profile is made.
func TestAProfileTakesTheNameAProjectHad(t *testing.T) {
	home := build(t, map[string]string{
		"profiles/ada.json": "{}\n",
		"ada/one.md":        "hello\n",
	})
	run(t, home, plain())
	has(t, home, "ada/profile.json", "ada/ada/local/one.md")
}

// A migration that stopped halfway is the only way this happens, and neither folder is worth
// losing to the other.
func TestKeepsBothWhereTheNameIsAlreadyTaken(t *testing.T) {
	home := build(t, map[string]string{
		"ada/profile.json":        "{}\n",
		"ada/notes/local/kept.md": "kept\n",
		"notes/moved.md":          "moved\n",
	})
	run(t, home, plain())
	has(t, home, "ada/notes/local/kept.md", "ada/notes-2/local/moved.md")
}

// The layout before a project held checkouts had the project folder be the checkout itself.
func TestLiftsTheCheckoutIntoLocalAndLeavesTheReposBesideIt(t *testing.T) {
	home := build(t, map[string]string{
		"ada/profile.json":                 "{}\n",
		"ada/notes/one.md":                 "hello\n",
		"ada/notes/.projects/x/local/a.md": "x\n",
	})
	run(t, home, plain())
	has(t, home, "ada/notes/local/one.md", "ada/notes/.repos/x/local/a.md")
	hasNo(t, home, "ada/notes/.projects/", "ada/notes/local/.repos/")
}

// A project already holding a checkout is left exactly as it is.
func TestLeavesAProjectThatAlreadyHasACheckout(t *testing.T) {
	home := build(t, map[string]string{
		"ada/profile.json":         "{}\n",
		"ada/notes/local/one.md":   "hello\n",
		"ada/notes/feature/one.md": "hello\n",
	})
	run(t, home, plain())
	has(t, home, "ada/notes/local/one.md", "ada/notes/feature/one.md")
	hasNo(t, home, "ada/notes/local/local/")
}

// Every repository the registry pointed at, moved into the project as the repo's own `local`.
func TestAdoptsEveryRepositoryTheRegistryPointedAt(t *testing.T) {
	home := build(t, map[string]string{
		"ada/profile.json":       "{}\n",
		"ada/notes/local/one.md": "hello\n",
	})
	// A repository anywhere on the disk, which is what the registry was for. Outside the home,
	// or it would be read as a project from the old layout and moved before this ran.
	elsewhere := filepath.Join(t.TempDir(), "api")
	if err := os.MkdirAll(elsewhere, 0o755); err != nil {
		t.Fatal(err)
	}
	if err := os.WriteFile(filepath.Join(elsewhere, "main.go"), []byte("package main\n"), 0o644); err != nil {
		t.Fatal(err)
	}
	registry := map[string]string{"api": elsewhere, "gone": filepath.Join(home, "nothing")}
	body, err := json.Marshal(registry)
	if err != nil {
		t.Fatal(err)
	}
	if err := os.MkdirAll(filepath.Join(home, "ada", "notes", ".repos"), 0o755); err != nil {
		t.Fatal(err)
	}
	if err := os.WriteFile(filepath.Join(home, "ada", "notes", ".repos", "projects.json"), body, 0o644); err != nil {
		t.Fatal(err)
	}

	run(t, home, plain())
	has(t, home, "ada/notes/.repos/api/local/main.go")
	// The registry the repos replaced is the one thing the migration deletes, and an entry with
	// nothing behind it leaves nothing behind.
	hasNo(t, home, "ada/notes/.repos/projects.json", "ada/notes/.repos/gone/")
}

// Tasks were called dreams, and the name was in the extension every one of them wore. Every
// checkout, since a branch has a copy.
func TestRenamesEveryDreamToATask(t *testing.T) {
	home := build(t, map[string]string{
		"ada/profile.json":                 "{}\n",
		"ada/notes/local/.tasks/one.dream": "{}\n",
		"ada/notes/local/deep/two.dream":   "{}\n",
		"ada/notes/feature/three.dream":    "{}\n",
		// A repo is somebody else's source and is never walked.
		"ada/notes/.repos/x/local/four.dream": "{}\n",
	})
	run(t, home, plain())
	has(t, home,
		"ada/notes/local/.tasks/one.task",
		"ada/notes/local/deep/two.task",
		"ada/notes/feature/three.task",
		"ada/notes/.repos/x/local/four.dream",
	)
	hasNo(t, home, "ada/notes/local/.tasks/one.dream")
}

// A dream whose task is already there is left alone rather than written over.
func TestLeavesADreamWhoseTaskIsAlreadyThere(t *testing.T) {
	home := build(t, map[string]string{
		"ada/profile.json":          "{}\n",
		"ada/notes/local/one.dream": "old\n",
		"ada/notes/local/one.task":  "new\n",
	})
	run(t, home, plain())
	body, err := os.ReadFile(filepath.Join(home, "ada", "notes", "local", "one.task"))
	if err != nil {
		t.Fatal(err)
	}
	if string(body) != "new\n" {
		t.Errorf("the task now says %q", body)
	}
	has(t, home, "ada/notes/local/one.dream")
}

// `attachments/` is dotted like every other folder the app owns, and `.skills/` moves under
// `.tools/` beside what a skill runs.
func TestRenamesTheTwoFoldersInsideACheckout(t *testing.T) {
	home := build(t, map[string]string{
		"ada/profile.json":                       "{}\n",
		"ada/notes/local/attachments/one.png":    "png\n",
		"ada/notes/local/.skills/hello/SKILL.md": "hi\n",
		"ada/notes/feature/attachments/two.png":  "png\n",
	})
	run(t, home, plain())
	has(t, home,
		"ada/notes/local/.attachments/one.png",
		"ada/notes/local/.tools/.skills/hello/SKILL.md",
		"ada/notes/feature/.attachments/two.png",
	)
	hasNo(t, home, "ada/notes/local/attachments/", "ada/notes/local/.skills/")
}

// Everything this machine filed under a project path, filed under the path it now has.
func TestRewritesTheConfigOntoThePathsThatMoved(t *testing.T) {
	home := build(t, map[string]string{
		"ada/profile.json": "{}\n",
		"notes/one.md":     "hello\n",
	})
	was := filepath.Join(home, "notes")
	now := filepath.Join(home, "ada", "notes")

	loaded := plain()
	loaded.Config.ProjectPath = &was
	loaded.Config.Checkouts = map[string]string{was: "feature"}
	loaded.Config.Repo = map[string]*string{was: nil}
	loaded.Config.RepoBranch = map[string]string{was + "#api": "trunk", "no-hash": "kept"}
	held := run(t, home, loaded)

	if held.ProjectPath == nil || *held.ProjectPath != now {
		t.Errorf("the open project is %v, want %s", held.ProjectPath, now)
	}
	if held.Profile == nil || *held.Profile != "ada" {
		t.Errorf("profile is %v", held.Profile)
	}
	if held.Checkouts[now] != "feature" {
		t.Errorf("checkouts are %v", held.Checkouts)
	}
	if _, said := held.Repo[now]; !said {
		t.Errorf("repo is %v", held.Repo)
	}
	if held.RepoBranch[now+"#api"] != "trunk" || held.RepoBranch["no-hash"] != "kept" {
		t.Errorf("repoBranch is %v", held.RepoBranch)
	}
}
